JS-Nerdery-Challenges/Rafael-David

RafoDev opened the pull request:
Solution of the challenge algorithms.

RafoDev committed:
feat(challenges): complete exercises 1 to 5

marceloev123 commented on JS-Algorithms/challenges.js:
```diff
@@ -13,7 +13,15 @@ Invoking "readableTime(3690)" should return "01:01:30" (HH:MM:SS)
+
+  const units = [hours, minutes, remainingSeconds];
+
+  const formattedUnits = units.map((unit) => unit.toString().padStart(2, '0'));
```
Nice way yo apply the needed unit transformation, I would prefer a better variable name like timeUnits instead of units but good work 👍

marceloev123 commented on JS-Algorithms/challenges.js:
```diff
@@ -13,7 +13,15 @@ Invoking "readableTime(3690)" should return "01:01:30" (HH:MM:SS)
+
+  const formattedUnits = units.map((unit) => unit.toString().padStart(2, '0'));
+
+  return formattedUnits.join(':');
```
you can also return this directly right? but I understand why you separate this to have more readability 💯 

return timeUnits.map((unit) => unit.toString().padStart(2, '0')).join(':');

marceloev123 commented on JS-Algorithms/challenges.js:
```diff
@@ -41,7 +49,16 @@ Invoking "circularArray(2)" should return "["Island", "Japan", "Israel", "German
 
 const circularArray = (index) => {
-  // YOUR CODE HERE...
+
```
What could happen if I send a negative number as an index here?

marceloev123 replied:
Which value will be appended if index is negative?

RafoDev replied:
It will append an `undefined` at the beginning, I forgot to make that check.

marceloev123 commented on JS-Algorithms/challenges.js:
```diff
@@ -70,7 +87,28 @@ The last 3 digits for the sum of powers from 1 to 10 is "317"
+    .reduce((acc, currNum) => (acc + currNum) % modBase, 0)
+    .toString();
+
+  const boundedSumOfPowers = sumOfPowers.slice(sumOfPowers.length - lastDigits);
```
You can use a negative value in the slice function and it would have the same result

```suggestion
  const boundedSumOfPowers = sumOfPowers.slice(-lastDigits);
```

marceloev123 commented on JS-Algorithms/challenges.js:
```diff
@@ -70,7 +87,28 @@ The last 3 digits for the sum of powers from 1 to 10 is "317"
 
 const ownPower = (number, lastDigits) => {
-  // YOUR CODE HERE...
+  const modBase = Math.pow(10, lastDigits + 1);
```
I like this approach to not overflow numbers and be precise in the calculations, you can also use a BigInt instead of it but good usage

marceloev123 commented on JS-Algorithms/challenges.js:
```diff
@@ -70,7 +87,28 @@ The last 3 digits for the sum of powers from 1 to 10 is "317"
+
+  const boundedSumOfPowers = sumOfPowers.slice(sumOfPowers.length - lastDigits);
+
+  return boundedSumOfPowers;
```
Good solution, I like the different approach maybe for future references bear in mind that some of the functions can be chained

```
const modBase = Math.pow(10, lastDigits + 1);
const sumOfPowers = Array.from({ length: number }, (_, idx) => {
    let poweredNumber = 1;
    for (let i = 0; i < idx + 1; i++) {
      poweredNumber = (poweredNumber * (idx + 1)) % modBase;
    }
    return poweredNumber;
  })
    .reduce((acc, currentNumber) => (acc + currentNumber) % modBase, 0)
    .toString();
    
   return sumOfPowers.slice(-lastDigits);
```

RafoDev replied:
At first, I thought about sending it like that, but I wasn't sure how readable it would be. I'll keep it in mind for next time.

marceloev123 commented on JS-Algorithms/challenges.js:
```diff
@@ -70,7 +87,28 @@ The last 3 digits for the sum of powers from 1 to 10 is "317"
+
+  const serie = Array.from({ length: number }, (_, idx) => idx + 1);
+
+  const poweredSerie = serie.map((num) => {    
```
I know why you declare this variable as num but maybe you can be more descriptive in your variables, don't be shy of using large naming, maybe `serieNumber` ?

marceloev123 commented on JS-Algorithms/challenges.js:
```diff
@@ -70,7 +87,28 @@ The last 3 digits for the sum of powers from 1 to 10 is "317"
+  });
+
+  const sumOfPowers = poweredSerie
+    .reduce((acc, currNum) => (acc + currNum) % modBase, 0)
```
same about naming conventions
```suggestion
    .reduce((acc, currentNumber) => (acc + currentNumber) % modBase, 0)
```

marceloev123 commented on JS-Algorithms/challenges.js:
```diff
@@ -95,7 +133,19 @@ Since 10! === 3628800 and you sum 3 + 6 + 2 + 8 + 8 + 0 + 0
+    factorialOfN *= i;
+  }
+
+  const digitsInfactorialOfN = [...factorialOfN.toString()].map(
```
Why spread operator is needed?

RafoDev replied:
Because `factorialOfN.toString()` returns a string, using the spread operator splits the string into individual characters and place them in an array, allowing me to apply a `map`.

marceloev123 commented on JS-Algorithms/challenges.js:
```diff
@@ -95,7 +133,19 @@ Since 10! === 3628800 and you sum 3 + 6 + 2 + 8 + 8 + 0 + 0
+
+  const sumOfDigits = digitsInfactorialOfN.reduce((acc, num) => acc + num, 0);
+
+  return sumOfDigits;
```
Clever solution you can apply the number cast directly to the reduce function instead of separate it in two steps though 

` const sumOfDigits = [...factorialOfN.toString()].reduce((acc, digit) => acc + Number(digit), 0);
`

marceloev123 commented on JS-Algorithms/challenges.js:
```diff
@@ -118,7 +168,21 @@ Because the 12th index in the Fibonacci sequence is 144, and 144 has three digit
+  let index = 2;
+
+  // f(n) = f(n-1) + f(n-2)  
+  let n1 = 0;
```
Clear variables definition good work, however same thing about naming you can use other variable names

n1 = previousNumber
n2 = currentNumber

I would encourage you to use more descriptive names for variables

marceloev123 commented on JS-Algorithms/challenges.js:
```diff
@@ -118,7 +168,21 @@ Because the 12th index in the Fibonacci sequence is 144, and 144 has three digit
+  // f(n) = f(n-1) + f(n-2)  
+  let n1 = 0;
+  let n2 = 1;
+  let fn = 1;
```
I think fn can be removed, you can use n2 variable and have the same result right?

RafoDev replied:
That's true, I hadn't noticed it.

marceloev123 reviewed (commented):
Good work on finishing your JS challenge 🎉 , I noticed that sometimes you tend to separate things in extra steps, I think that you did it intentionally for code readability, however bear in mind that extra steps sometimes are unnecessary, I would like you to improve variables naming for future exercise too, good solutions so far, I think you have a good understanding when to use declarative functions

RafoDev committed:
refactor(challenges): improve readability and efficiency

- Suggested more descriptive variable names for better clarity.
- Removed redundant lines by chaining functions.
